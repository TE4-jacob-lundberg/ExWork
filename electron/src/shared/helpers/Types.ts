export interface ILink {
  url: string;
  label: string;
  type: string;
}

export interface IGameFormFields {
  title: string;
  abbreviation: string;
  image: string; 
  xAxis: string;
  yAxis: string;
  fileNames: string[];
}

export interface IGameFormErrors { [id: string]: string; }

export interface IGame {
  id: string;
  title: string;
  abbreviation: string | null;
  image: IGameImage;
  links: ILink[];
  fileNames: string[];
}

export interface IGameImage {
  url: string;
  bannerPos: IPosition;
}

export interface IPosition {
  x: string;
  y: string;
}
