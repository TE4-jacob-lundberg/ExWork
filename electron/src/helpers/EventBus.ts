/* eslint-disable @typescript-eslint/no-explicit-any */
const EventBus: IEventBus = {
  events: {},
  dispatch: function (event: string, data: any): void {
    if (!this.events[event]) return;
    this.events[event].forEach((callBack: Function) => callBack(data));
  },
  subscribe: function (event: string, callBack: Function): void {
    if (!this.events[event]) this.events[event] = [];
    this.events[event].push(callBack);
  },
};

interface IEventBus {
  events: {[id: string]: Function[]};
  dispatch: (event: string, data: any) => void;
  subscribe: (event: string, callBack: Function) => void;
}

export { EventBus };
