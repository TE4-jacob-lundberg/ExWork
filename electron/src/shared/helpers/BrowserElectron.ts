// eslint-disable-next-line @typescript-eslint/no-explicit-any
export let electron: any;

if (process.env.ENVIRONMENT === 'test') electron = require('electron');
else electron = window.require('electron');
