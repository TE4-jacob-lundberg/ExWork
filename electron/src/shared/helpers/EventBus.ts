/* eslint-disable @typescript-eslint/no-explicit-any */
interface IEventBus {
  events: {[id: string]: Function[]};
  dispatch: (event: string, data: any) => void;
  subscribe: (event: string, callBack: Function) => void;
}

const EventBus: IEventBus = {
  events: {},
  dispatch: function (event: string, data: any): void {
    if (!this.events[event]) return;
    this.events[event].forEach((callBack: Function) => callBack(data));
  },
  subscribe: function (event: string, callBack: Function): void {
    if (!this.events[event]) this.events[event] = [];
    this.events[event].push(callBack);
  },
};

export { EventBus };
