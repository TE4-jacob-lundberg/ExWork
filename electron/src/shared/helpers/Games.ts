import { standardGames, IGame } from '../constants/standardGames';

export class Games {

  public static all(): IGame[] {
    if (!localStorage.getItem('games')) localStorage.setItem('games', JSON.stringify(standardGames));

    return JSON.parse(localStorage.getItem('games')!) || [];
  }

  public static addNew(game: IGame): string {
    const savedGames: IGame[] = JSON.parse(localStorage.getItem('games')!);
    if (savedGames.some(sGame => sGame.id === game.id)) return 'DUPLICATE_ID';
    
    localStorage.setItem('games', JSON.stringify(savedGames.concat(game)));
    return '';
  }
}
