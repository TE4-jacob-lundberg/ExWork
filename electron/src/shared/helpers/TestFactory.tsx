/* eslint-disable @typescript-eslint/no-explicit-any */
import React from 'react';
import { render, RenderResult } from '@testing-library/react';
import { Router } from 'react-router-dom';
import { createBrowserHistory } from 'history';

export function factory(component: JSX.Element): RenderResult {
  return render(component);
}

export function routerFactory(component: JSX.Element, options: any = { history: createBrowserHistory() }): RenderResult {
  return factory(
    <Router history={options.history}>
      {component}
    </Router>
  );
}
