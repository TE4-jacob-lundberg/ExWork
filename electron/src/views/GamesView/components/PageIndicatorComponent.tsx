import React from 'react';
import Styled from '@emotion/styled';

type ContainerProps = {
  amount: number;
}

const ContainerStyled = Styled.div<ContainerProps>`
  position: absolute;
  display: grid;
  grid-template-columns: repeat(${(props): string => props.amount.toString()}, 1fr);
  bottom: -36px;
`;

type IndicatorProps = {
  active: boolean;
}

const IndicatorStyled = Styled.div<IndicatorProps>`
  height: 16px;
  width: 16px;
  border-radius: 999px;
  background: ${(props): string => props.active ? '#222' : '#666'};
  margin: 8px;
`;

interface Props {
  totalPages: number;
  currentPage: number;
  onSelect: (pageNumber: number) => void;
}

const PageIndicatorComponent: React.FC<Props> = function (props: Props) {
  return (
    <ContainerStyled amount={props.totalPages}>
      {Array(props.totalPages).fill(null).map((val, index) => 
        <IndicatorStyled 
          key={index} 
          active={props.currentPage === index} 
          onClick={(): void => props.onSelect(index)}
          data-testid="indicator"
        />
      )}
    </ContainerStyled>
  );
};

export default PageIndicatorComponent;
