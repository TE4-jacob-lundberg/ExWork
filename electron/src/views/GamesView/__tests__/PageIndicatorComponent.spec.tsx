import React from 'react';
import { fireEvent } from '@testing-library/react';

import { factory } from '../../../shared/helpers/TestFactory';
import PageIndicatorComponent from '../components/PageIndicatorComponent';

const mockedSelect = jest.fn();

const testComponent = <PageIndicatorComponent 
  totalPages={2}
  currentPage={0}
  onSelect={mockedSelect}
/>;

describe('PageIndicatorComponent.tsx', () => {
  it('renders correctly', () => {
    const { container } = factory(testComponent);
    
    expect(container).toMatchSnapshot();
  });

  it('updates whenever the page changes', () => {
    const { container, rerender } = factory(testComponent);

    expect(container).toMatchSnapshot();

    rerender(<PageIndicatorComponent 
      totalPages={2}
      currentPage={1}
      onSelect={jest.fn()}
    />);

    expect(container).toMatchSnapshot();
  });

  it('returns the new page on select', () => {
    const { getAllByTestId } = factory(testComponent );

    fireEvent.click(getAllByTestId('indicator')[1]);

    expect(mockedSelect).toHaveBeenCalledWith(1);
  });
});
