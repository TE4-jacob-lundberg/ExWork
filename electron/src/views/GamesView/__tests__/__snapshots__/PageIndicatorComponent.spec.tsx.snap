// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PageIndicatorComponent.tsx renders correctly 1`] = `
<div>
  <div
    class="css-i1e77v"
  >
    <div
      class="css-ld341p"
      data-testid="indicator"
    />
    <div
      class="css-1sr9bqy"
      data-testid="indicator"
    />
  </div>
</div>
`;

exports[`PageIndicatorComponent.tsx updates whenever the page changes 1`] = `
<div>
  <div
    class="css-i1e77v"
  >
    <div
      class="css-ld341p"
      data-testid="indicator"
    />
    <div
      class="css-1sr9bqy"
      data-testid="indicator"
    />
  </div>
</div>
`;

exports[`PageIndicatorComponent.tsx updates whenever the page changes 2`] = `
<div>
  <div
    class="css-i1e77v"
  >
    <div
      class="css-1sr9bqy"
      data-testid="indicator"
    />
    <div
      class="css-ld341p"
      data-testid="indicator"
    />
  </div>
</div>
`;
